contracts/upgradeable_contracts/factories/HomeBridgeFactory.sol

```diff
@@ -0,0 +1,216 @@
+pragma solidity 0.4.24;
+
+import "../../IBridgeValidators.sol";
+import "../../IHomeBridge.sol";
+import "../../upgradeability/EternalStorageProxy.sol";
+import "../../upgradeability/EternalStorage.sol";
+import "../../ERC677BridgeToken.sol";
+import "../EternalOwnable.sol";
+
+contract HomeBridgeFactory is EternalStorage, EternalOwnable {
+
+  event HomeBridgeDeployed(address indexed _homeBridge, address indexed _homeValidators, address indexed _token, uint256 _blockNumber);
+
+  function getBridgeFactoryVersion() public pure returns(uint64 major, uint64 minor, uint64 patch) {
+    return (2, 2, 0);
+  }
+
+  function initialize(address _owner,
+      address _bridgeValidatorsImplementation,
+      uint256 _requiredSignatures,
+      address[] _initialValidators,
+      address _bridgeValidatorsOwner,
+      address _homeBridgeErcToErcImplementation,
+      uint256 _requiredBlockConfirmations,
+      uint256 _gasPrice,
+      uint256 _dailyLimit,
+      uint256 _maxPerTx,
+      uint256 _minPerTx,
+      uint256 _foreignDailyLimit,
+      uint256 _foreignMaxPerTx,
+      address _homeBridgeOwner,
+      address _homeProxyOwner) public {
+    
+
+    require(_owner != address(0));
+    require(_bridgeValidatorsImplementation != address(0));
+    require(_requiredSignatures >= 1);
+    require(_bridgeValidatorsOwner != address(0));
+    require(_homeBridgeErcToErcImplementation != address(0));
+    require(_gasPrice > 0);
+    require(_requiredBlockConfirmations > 0);
+    require(_minPerTx > 0 && _maxPerTx > _minPerTx && _dailyLimit > _maxPerTx);
+    require(_foreignMaxPerTx < _foreignDailyLimit);
+    require(_homeBridgeOwner != address(0));
+    require(_homeProxyOwner != address(0));
+
+    setOwner(_owner);
+    setBridgeValidatorsImplementation(_bridgeValidatorsImplementation);
```

Comment:
While testing, I found an issue on how variables are set on `initialize` method. Calling methods that has the `onlyOwner` modifier will only work if the deployment account(`DEPLOYMENT_ACCOUNT_PRIVATE_KEY`) is the same as `HOME_FACTORY_OWNER`. This may not always be the case, setting multisignature wallets as owner of contracts is a common pattern.

To avoid this issue, on `initialize` method of other bridge modes we set the values directly:
https://github.com/poanetwork/poa-bridge-contracts/blob/46151629adb3dd20e6fb47d4ba91b88c6d388c4f/contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol#L38-L46

Another option could be to define one internal method for every parameter like the following:
```solidity
  function _setBridgeValidatorsImplementation(address _bridgeValidatorsImplementation) internal {
    addressStorage[keccak256(abi.encodePacked("bridgeValidatorsImplementation"))] = _bridgeValidatorsImplementation;
  }
```
And call it from `setBridgeValidatorsImplementation` and `initialize`.

The same issue should be addressed on `ForeignBridgeFactory`

Reply:
I understand the problem you mentioned, I will suggest (and implement) a third solution:
First set the owner as the `msg.sender` so that I could call all the setters from within the `initialize` method and at the end of the method, I will set the owner to be the `_owner` parameter.